### PR TITLE
8282392: [zero] Build broken on AArch64

### PR DESCRIPTION
**Pull request by a74nh**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282392](https://bugs.openjdk.java.net/browse/JDK-8282392): [zero] Build broken on AArch64

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to d5952abcdbdcddd3311e77b870ea0723cb91ae27
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7633/head:pull/7633` \
`$ git checkout pull/7633`

Update a local copy of the PR: \
`$ git checkout pull/7633` \
`$ git pull https://git.openjdk.java.net/jdk pull/7633/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7633`

View PR using the GUI difftool: \
`$ git pr show -t 7633`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7633.diff">https://git.openjdk.java.net/jdk/pull/7633.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/macros.hpp [hunk 1]
```diff
@@ -537,6 +537,14 @@
 #define NOT_AARCH64(code) code
 #endif
 
+#ifdef TARGET_ARCH_aarch64
+#define AARCH64_PORT_ONLY(code) code
+#define NOT_AARCH64_PORT_ONLY(code)
+#else
```
Comment (theRealAph) [inline src/hotspot/share/utilities/macros.hpp:543]: I don't think we need `NOT_AARCH64_PORT_ONLY`, and it's too confusing. Otherwise OK.

Comment (a74nh) [reply src/hotspot/share/utilities/macros.hpp:543]: Agreed. Only added it to keep with the style of the file. Will remove.

**Review by theRealAph — APPROVED** [review]

**Review by shipilev — COMMENTED** [review]
I think it is confusing to have `AARCH64_PORT_ONLY` defines, to be honest. In the similar cases for X86, we just additionally protect these blocks with !ZERO. Something like:

```
#if defined(AARCH64) && !defined(ZERO)
ret_pc = pauth_strip_pointer(ret_pc);
#endif
```

**Comment by theRealAph** [comment]
> I think it is confusing to have `AARCH64_PORT_ONLY` defines, to be honest. In the similar cases for X86, we just additionally protect these blocks with !ZERO. Something like:

That's what we looked at and it was more of a mess, IMO. In the end it's a judgment call which to have, and I've seen this kind of mistake, where a particular port is confused with a particular CPU, enough times that I think this is OK; YMMV.

**Comment by shipilev** [comment]
> That's what we looked at and it was more of a mess, IMO. In the end it's a judgment call which to have, and I've seen this kind of mistake, where a particular port is confused with a particular CPU, enough times that I think this is OK; YMMV.

From the perspective of Zero maintenance, having the Zero-specific workarounds explicitly doing `!ZERO` is cleaner. This mess is mostly Zero-s problem with idenitifying itself as CPU. So, in my mind, there is little reason to accommodate that problem with "port" defines.

**Comment by theRealAph** [comment]
> > That's what we looked at and it was more of a mess, IMO. In the end it's a judgment call which to have, and I've seen this kind of mistake, where a particular port is confused with a particular CPU, enough times that I think this is OK; YMMV.
> 
> From the perspective of Zero maintenance, having the Zero-specific workarounds explicitly doing `!ZERO` is cleaner. This mess is mostly Zero-s problem with idenitifying itself as CPU. So, in my mind, there is little reason to accommodate that problem with "port" defines.

I think I understand your point, but IMO it's almost always easier to understand language which says what something is than what it isn't, and a simple name than a boolean expression. And that is more important, I believe.
Having said that, if you insist that flagging this up as a Zero-specific workaround with `!ZERO` is really important I will give way to your preference. (I don't think it is: I think we should flag this code as port-specific, not CPU-specific. But mostly I just want this patch pushed.)

**Comment by a74nh** [comment]
My only issue with that is that:
AARCH64_PORT_ONLY(some_function());

becomes:
#if defined(AARCH64) && !defined(ZERO)
some_function();
#endif

Which is a little uglier.

How about defining the macro something like:
#if defined(AARCH64) && !defined(ZERO)
#define AARCH64_NOT_ZERO(code) code


(ultimately, I'm happy with any of the above)

**Review by shipilev — APPROVED** [review]
Fine, let's do it in this form.

**Comment by dholmes-ora** [comment]
Sorry I missed this but this is stylistically awful IMHO. What is AARCH64_PORT_ONLY supposed to mean? IIUC it really means AARCH64_NATIVE_PORT_ONLY or AARCH64_NOT_ZERO_ONLY. I would much rather have seen @shipilev request to use a combination of CPU and ZERO to get this right - as used everywhere else when dealing with ZERO.
And we don't use TARGET_ARCH_x anywhere else in the code other than to define the macros.

**Comment by theRealAph** [comment]
> Sorry I missed this but this is stylistically awful IMHO. What is AARCH64_PORT_ONLY supposed to mean? IIUC it really means AARCH64_NATIVE_PORT_ONLY or AARCH64_NOT_ZERO_ONLY. I would much rather have seen @shipilev request to use a combination of CPU and ZERO to get this right - as used everywhere else when dealing with ZERO. And we don't use TARGET_ARCH_x anywhere else in the code other than to define the macros.

I've seen this problem repeating over the years, caused by a failure to distinguish between an ifdef for the properties of a particular CPU and one for the implementation of a particular port. Courtesy of Zero, there is not a 1:1 mapping between these, and (judging by the problems we've seen) it is a frequent cause of confusion. Usually we catch the bugs before push, but not always.

In my opinion that this is not merely a matter of people making mistakes, but of a style that is confusing, and will continue to be.

**Comment by dholmes-ora** [comment]
I prefer adding the ifndef ZERO to guard these cases than this singular attempt to do something completely different.

**Comment by magicus** [comment]
@theRealAph Hear, hear!

Zero is causing a conceptual mess the way it is currently "injected" into hotspot. This mess also spills over into the build system.

The core problem is that Zero claims to be a "CPU", which it clearly is not. And then a lot of workarounds needs to be installed everywhere to compensate for this confusion. A better solution would be to go back to CPU meaning just CPU, and introduce a second dimension along which to determine zero/non-zero. We might have to bikeshed for a while to find a good name for this second dimension (I can't come up with any ideal names straight away, anyway. INTERPRETED_ONLY vs COMPILED? ZERO vs NON_ZERO? Or ZERO vs ONE? :-))

Doing this would simplify the concepts and logic behind both the Hotspot code and the build system, and eliminate a class of errors that keep popping up all the time, due to this confusing design decision.

**Comment by dholmes-ora** [comment]
Zero is a CPU-agnostic interpreter build, but our builds are inherently CPU-based, so Zero has to represent that "don't care" CPU because it has to replace some CPU specific code with Zero's C code. But then you have to build other CPU-specific parts of the JDK even when using Zero. Hence the approach of using CPU ifdefs combined with a check for Zero.
Yes it is awkward and confusing and mistakes can creep in. But I don't think you will come up with anything better for this situation. And the AARCH64_PORT_ONLY is certainly not better IMO.

**Comment by theRealAph** [comment]
> Zero is a CPU-agnostic interpreter build, but our builds are inherently CPU-based, so Zero has to represent that "don't care" CPU because it has to replace some CPU specific code with Zero's C code. But then you have to build other CPU-specific parts of the JDK even when using Zero. Hence the approach of using CPU ifdefs combined with a check for Zero. Yes it is awkward and confusing and mistakes can creep in. But I don't think you will come up with anything better for this situation. And the AARCH64_PORT_ONLY is certainly not better IMO.

Can you explain why it's not better? I don't want to waste your time by prolonging discussions unnecessarily, but it seems to me that this nomenclature conveys exactly what we mean: not so much running on a particular CPU, but a particular port.
There's a matrix of possible combinations of instruction set architecture and port, with the ports ranging from highly-tuned and customized to completely portable, with space in between.

**Comment by dholmes-ora** [comment]
@theRealAph the word "port" may have a very specific meaning to you such that this macro conveys what you expect, but it doesn't to me. You have to know what is not considered part of a "port" when building Aarch64 - and that is simply when building Zero. So to me Zero should be part of any conditionalisation. e.g. (if macros work this way)

`NOT_ZERO(AARCH64_ONLY(ret_pc = pauth_strip_pointer(ret_pc);))`

or the ifdefs @shipilev  suggested.

**Pull request closed without merging**